### PR TITLE
ci(mutation-tests): pass per-folder test-case-filter to bypass xUnit v3 perTest bug

### LINKED ISSUE #1027: [INFRA] Pass per-folder test-case-filter to Stryker to bypass xUnit v3 perTest bug
## Category

- [x] CI/CD Workflow (GitHub Actions)
- [x] Test Infrastructure (fixtures, runners, parallelization)

## Description

The Mutation Tests workflow currently runs each mutant against the full ~23,000-test suite because Stryker.NET cannot use `perTest` coverage analysis with xUnit v3 (upstream bug tracked in #1026). This makes kill results semi-stochastic and adding targeted tests yields no measurable improvement — see the empirical evidence in the closing comment of #962.

We can bypass the upstream bug by supplying Stryker with a **manual test-to-source mapping** based on Encina's namespace conventions. Test projects mirror source structure: `src/Encina/Sharding/Migrations/Strategies/*.cs` is tested exclusively by `tests/Encina.UnitTests/Core/Sharding/Migrations/Strategies/*Tests.cs`. So when Stryker is asked to mutate a given folder, we can narrow the `test-case-filter` to only the tests whose namespace mirrors that folder. Each mutant then runs the ~20-50 relevant tests instead of 23,000.

## Component Affected

- **File(s)**:
  - `.github/workflows/mutation-tests.yml` — the `Select mutation scope` step needs to also emit a `TEST_FILTER` output
  - `.github/scripts/run-stryker.cs` — forward the `--test-case-filter` argument when provided
- **Environment**: GitHub Actions `ubuntu-latest`, Stryker.NET 4.14.0, xUnit v3

## Current Behavior

- `custom_scope` input (added in #962) narrows the `--mutate` glob to one folder
- `test-case-filter` in `stryker-config.json` is fixed to the three declared test projects' namespaces (`FullyQualifiedName~Encina.UnitTests|...~Encina.PropertyTests|...~Encina.ContractTests`) — every run ignores that folder's actual scope
- With `coverage-analysis: "off"`, Stryker runs every test against every mutant, which both slows the run (~3 min/mutant × ~60 mutants ≈ 3 h) and makes kills non-deterministic
- Run [24584286635](https://github.com/dlrivada/Encina/actions/runs/24584286635): 1 killed / 58 survived despite 26 mutation-targeted tests being in place

## Expected Behavior

- The `Select mutation scope` step also emits a `TEST_FILTER` output derived from the mutation scope:
  - `**/Sharding/Migrations/Strategies/*.cs` → `FullyQualifiedName~Sharding.Migrations.Strategies`
  - `**/Pipeline/Behaviors/*.cs` → `FullyQualifiedName~Pipeline.Behaviors`
  - ... and so on for each of the 17 rotation entries
- The `Run mutation tests` step passes `--test-case-filter "$TEST_FILTER"` to Stryker, overriding the config value
- Each mutant runs only tests from the matching namespace → ~20-50 per mutant instead of 23,000
- Per-mutant time drops from ~3 min to ~5-10 s
- A folder run completes in ~5-10 min instead of ~3 h
- Kills become deterministic because only tests that actually touch the mutated code path execute

## Impact

- **Developer Experience**: contributors adding mutation-targeted tests see immediate kill-count improvements
- **CI/CD Time**: per-folder run drops from ~3 h to ~5-10 min — ~20× reduction in runner time
- **Reliability**: kill count becomes a stable signal of test quality (no more "3 → 1" regressions from upstream version changes)
- **Unlocks #1027** (matrix-parallel full-project coverage)

### Phase 1: Folder-to-filter map in the workflow

- [ ] Extend the `FOLDERS` array in `.github/workflows/mutation-tests.yml` into a parallel structure that pairs each scope glob with its test filter. Options: two arrays (`FOLDERS` + `FILTERS`) indexed together, or a single associative array.
- [ ] Derive the filter from the scope glob automatically where possible — `**/Sharding/Migrations/Strategies/*.cs` ↔ `Sharding.Migrations.Strategies` is a clean mapping. Edge cases (`**/Core/*.cs` is ambiguous) can be hand-mapped.
- [ ] Emit the filter as a step output alongside the scope.

### Phase 2: Forward filter to Stryker

- [ ] Modify the `Run mutation tests` step to pass `--test-case-filter "$FILTER"` to `run-stryker.cs` when the filter is non-empty.
- [ ] `run-stryker.cs` already forwards unknown args to Stryker; verify no change needed there.

### Phase 3: Validate against the Strategies folder

- [ ] Trigger the workflow with `custom_scope='**/Sharding/Migrations/Strategies/*.cs'`. Expected: kill count jumps from 1 → somewhere in the 20-40 range (of 59 mutants), matching what the 26 targeted tests were designed to catch. Run should complete in ~5-10 min instead of ~3 h.

### Phase 4: Document

- [ ] Update `docs/testing/mutation-measurement-methodology.md`: new section "Per-folder test filter" explaining the manual mapping and its relationship to the upstream perTest bug.
- [ ] Update `CLAUDE.md` "Mutation Testing System" section.

## Risks

- **False negatives**: a contract test in a different namespace that would legitimately kill a mutant in the target folder won't run. For the current 17 rotation folders, source-to-test mapping is tight (dedicated `*Tests.cs` per source file), so the risk is small. If encountered, the fix is to broaden the filter for that specific folder.
- **Namespace drift**: if test namespaces get renamed without updating the mapping, kills silently disappear. Mitigation: the methodology doc documents the mapping convention; reviewers should catch mismatches.

## Related Issues

- #957 — Mutation workflow stabilization (closed)
- #962 — Scope widening + cited-by (closed)
- #1026 — Blocked-upstream: Stryker.NET xUnit v3 perTest bug (this issue is the workaround)
- Unlocks #1027 — Matrix-parallel full-project mutation coverage

## Repro (baseline, before fix)

```bash
gh workflow run "Mutation Tests" --ref main \
  -f custom_scope='**/Sharding/Migrations/Strategies/*.cs'
# Wait ~3h. Observe: 1 killed / 58 survived.
```

### PR DESCRIPTION
**Pull request by dlrivada**
## Summary

- Pairs each rotation folder in `mutation-tests.yml` with a test-namespace substring and forwards it to Stryker via `--test-case-filter "FullyQualifiedName~<substring>"`. Each mutant now runs the ~20–500 tests that mirror the mutated folder's namespace instead of the full ~23,000-test suite.
- `custom_scope` dispatch input reuses the paired filter automatically when the scope matches a rotation entry.
- Documents the mechanism and its edge cases (root-namespace folders, non-recursive globs, namespace drift risk) in the methodology doc, with a short pointer added to `CLAUDE.md`.

Closes #1027. Unblocks deterministic kill counts for the mutation-targeted tests added in #962.

## Why

Upstream xUnit v3 + Stryker `perTest` bug ([stryker-net#3117](https://github.com/stryker-mutator/stryker-net/issues/3117)) forces `AllTests` mode, so every mutant runs ~23,000 tests. That is both slow (~3 min/mutant × ~60 mutants ≈ 3 h) and stochastic — targeted tests get drowned in noise. `--test-case-filter` overrides `test-case-filter` in `stryker-config.json` at the VSTest level, so Stryker's missing `perTest` coverage is no longer on the critical path.

## Mapping design

| Source folder | Filter substring |
|---|---|
| `**/Sharding/Migrations/Strategies/*.cs` | `Sharding.Migrations.Strategies` |
| `**/Dispatchers/Strategies/*.cs` | `Dispatchers.Strategies` |
| `**/Modules/Isolation/*.cs` | `Modules.Isolation` |
| `**/Core/*.cs` (root ns) | `Encina.UnitTests.Core\|Encina.ContractTests.Core` |
| … see `FILTERS` array in the workflow for all 17 entries | |

The rule: use the source-folder dotted path as the substring where source has a named namespace; fall back to the explicit test-folder path when source is in `namespace Encina;` (root). Broader is always safe (extra tests are fine); narrower would risk false negatives.

## Test plan

- [ ] Trigger workflow with `custom_scope='**/Sharding/Migrations/Strategies/*.cs'`
- [ ] Confirm runtime drops from ~3 h to ~5–10 min
- [ ] Confirm kill count jumps from 1 → 20–40 range (26 targeted tests from #962 should now fire)
- [ ] Verify `filter` output is emitted correctly in the `Select mutation scope` step log
- [ ] Confirm empty `filter` (e.g. `full_mode: true`) still falls back to `stryker-config.json` default

## Notas de la versión

* **Documentation**
  * Documentación mejorada del sistema de pruebas de mutación con nueva sección sobre filtrado de pruebas por carpeta.
  * Actualización de la guía de ejecución con detalles sobre la optimización de selección de pruebas.

* **Chores**
  * Optimización del flujo de trabajo de pruebas de mutación para mejor enfoque en casos de prueba específicos.

### CODE REVIEW

File: .github/workflows/mutation-tests.yml [hunk 1]
```diff
@@ -137,10 +153,37 @@ jobs:
...
+            "Sharding.Routing"
+            "Sharding.ReplicaSelection"
+            "Sharding.Execution"
+            "Sharding.Shadow"
+            "Sharding.Colocation"
+            "Sharding.TimeBased"
+            "Sharding.Resharding"
+            "Sharding.Diagnostics"
+            "Modules.Isolation"
+            "Encina.UnitTests.Results|Encina.PropertyTests.Results"
```
[the remark this answers is not in the document]
Comment (dlrivada) [reply .github/workflows/mutation-tests.yml]: Good catch — confirmed. Tests for the Results source files (`EitherHelpers`, `EncinaErrors`, `NullFunctionalFailureDetector`) don't live in any `Results`-named namespace, so the original filter would have selected zero tests and all Results mutants would silently survive. Set the entry to an empty string in 009a2aef; empty falls back to the `stryker-config.json` default (all three Stryker test projects).

File: .github/workflows/mutation-tests.yml [hunk 2]
```diff
@@ -150,32 +193,55 @@ jobs:
             WEEK=$(date -u +%V)
             IDX=$(( 10#$WEEK % 17 ))
             SCOPE="${FOLDERS[$IDX]}"
+            FILTER="${FILTERS[$IDX]}"
```
[the remark this answers is not in the document]
Comment (dlrivada) [reply .github/workflows/mutation-tests.yml:208]: Agreed. Replaced the hard-coded `% 17` with `% ${#FOLDERS[@]}` and added an array-length assertion that fails fast when `FOLDERS` and `FILTERS` disagree (see 009a2aef). The mismatch would have been silent — a scope paired with the wrong filter — so a hard error is the right behavior.

File: .github/workflows/mutation-tests.yml [hunk 3]
```diff
@@ -150,32 +193,55 @@ jobs:
...
+            for part in "${FILTER_PARTS[@]}"; do
+              if [ -z "$TCF" ]; then
+                TCF="FullyQualifiedName~$part"
+              else
+                TCF="$TCF|FullyQualifiedName~$part"
+              fi
+            done
+            FILTER_ARGS=(--test-case-filter "$TCF")
+            echo "Test-case filter: $TCF"
+          fi
```
[the remark this answers is not in the document]
Comment (dlrivada) [reply .github/workflows/mutation-tests.yml]: Correct, and thanks for the web search — this was the critical finding in the review. Verified locally by running `dotnet stryker --help` against 4.14.0: no `--test-case-filter` option is listed. My initial CLI forwarding would have been silently ignored and every mutant would have kept running the full ~23k-test suite.

Fixed in 009a2aef by patching `.github/stryker-config.json` in place via `jq` before invoking Stryker (the `test-case-filter` property is the only place Stryker reads this setting from, per the docs and [stryker-net#3091](https://github.com/stryker-mutator/stryker-net/issues/3091)). CI runners are ephemeral so the in-place edit doesn't leak, and `run-stryker.cs` already points at `.github/stryker-config.json`, so no script changes were needed. Methodology doc and CLAUDE.md updated to reflect the actual mechanism.
